src/PeNet/Header/Resource/Icon.cs

```diff
@@ -40,132 +41,69 @@ public Icon(IRawFile peFile, long offset, uint size, uint id, Resources parent)
         /// <summary>
         ///     Byte span of the icon image.
         /// </summary>
-        public Span<byte> AsSpan()
+        public Span<byte> AsRawSpan()
         {
             return PeFile.AsSpan(Offset, Size);
         }
 
+        private bool IsPng => AsRawSpan().Slice(0, 8).SequenceEqual(PNGHeader);
+        private bool IsIco => !IsPng && AssociatedGroupIconDirectoryEntry is not null && !AsRawSpan().IsEmpty;
+
         /// <summary>
         ///     Adding .ICO-Header to the bytes of the icon image.
         /// </summary>
         /// <returns>Bytes of the icon image as .ICO file.</returns>
-        public byte[]? AsICO()
+        public byte[]? AsIco()
         {
-            if (AssociatedGroupIconDirectoryEntry == null) return null;
-            var iconBytes = AsSpan().ToArray();
-            if (iconBytes == null) return null;
+            var raw = AsRawSpan();
+            if (IsPng) return raw.ToArray(); // TODO: CFF does not use an additional header for .PNG. But this would not break anything.
+            if (!IsIco) return null;
 
-            if (iconBytes.Take(8).SequenceEqual(PNGHeader)) return iconBytes; // TODO: CFF does not use an additional header for .PNG. But this would not break anything.
+            var icoBytes = new byte[IcoHeaderSize + raw.Length];
+            var icoHeader = IcoHeader();
 
-            var iconICOBytes = new byte[ICOHeaderSize + iconBytes.Length];
-            SetICOHeader(iconICOBytes);
-            
-            iconBytes.CopyTo(iconICOBytes, ICOHeaderSize);
-            return iconICOBytes;
+            icoHeader.AsSpan().CopyTo(new Span<byte>(icoBytes, 0, (int) IcoHeaderSize));
```

Comment:
Nit: Can we try to write this a little more intuitively?

Reply:
Addressed this as well in the latest commit